frappe/__init__.py

```diff
@@ -1031,15 +1031,21 @@ def _respond(doc, from_redis=False):
 		if doc := cache().hget("document_cache", key):
 			return _respond(doc, True)
 
-	# Not found in local/redis, fetch from DB and store in cache
+	# Not found in local/redis, fetch from DB
 	doc = get_doc(*args, **kwargs)
 
-	# Store in redis cache
-	key = get_document_cache_key(doc.doctype, doc.name)
+	# Store in cache
+	if not key:
+		key = get_document_cache_key(doc.doctype, doc.name)
 
 	local.document_cache[key] = doc
-	# Avoid setting in local.cache since there's separate cache
-	cache().hset("document_cache", key, doc.as_dict(), cache_locally=False)
+
+	# Avoid setting in local.cache since we're already using local.document_cache above
+	# Try pickling the doc object as-is first, else fallback to doc.as_dict()
+	try:
+		cache().hset("document_cache", key, doc, cache_locally=False)
+	except Exception:
```

Comment:
Pickling the `Document` directly works for most DocTypes, except for **Quickbooks Migrator**.

can't use `pickle.PicklingError` here because it sometimes throws other errors, e.g. [`AttributeError`]()